parsec/mca/device/cuda/device_cuda_module.c

```diff
@@ -2781,4 +2877,91 @@ parsec_cuda_kernel_scheduler( parsec_execution_stream_t *es,
     return PARSEC_HOOK_RETURN_DISABLE;
 }
 
-#endif /* PARSEC_HAVE_DEV_CUDA_SUPPORT */
+/**
+ * Co Manager thread in charge of completing the tasks in stead of the
+ * manager itself. While completing the tasks, if any ready tasks are discovered
+ * they will be given back directly to the manager and will not go through the
+ * scheduler
+ */
+parsec_hook_return_t
+parsec_cuda_co_manager( parsec_execution_stream_t *es, parsec_device_gpu_module_t* gpu_device )
+{
+    int rc = 0;
+    parsec_task_t* task = NULL;
+    parsec_gpu_task_t *gpu_task = NULL;
+    parsec_list_t *gpu_tasks_to_free = NULL;
+    (void)es;
+
+    if( gpu_device->co_manager_mutex > 0 )
+    {
+        return PARSEC_HOOK_RETURN_ASYNC;
+    }
+    else
+    {
+        rc = gpu_device->co_manager_mutex;
+        if( !parsec_atomic_cas_int32( &gpu_device->co_manager_mutex, rc, rc+1 ) )
+        {
+            return PARSEC_HOOK_RETURN_ASYNC;
+        }
+    }
+
+    PARSEC_DEBUG_VERBOSE(2, parsec_gpu_output_stream,"GPU[%s]: Thread %d Entering GPU co-management at %s:%d",
+                         gpu_device->super.name, es->th_id, __FILE__, __LINE__);
+
+    /**
+     * Thread-local variable to identify that this thread is the co-manager
+     * The value allows to differentiate between the different co-managers (debug output mostly)
+     */
+    parsec_device_module_t** co_manager_tls_val = malloc(sizeof(parsec_device_module_t*));
+    *co_manager_tls_val = &(gpu_device->super);
+    PARSEC_TLS_KEY_CREATE(co_manager_tls);
+    PARSEC_TLS_SET_SPECIFIC(co_manager_tls, co_manager_tls_val);
+
+    gpu_tasks_to_free = PARSEC_OBJ_NEW(parsec_list_t);
+
+    /**
+     * The co-manager can be created before any task is to be completed, so wait while mutex > 0
+     * then complete the tasks.
+     */
+    while( gpu_device->mutex > 0 || gpu_device->complete_mutex > 0 )
+    {
+        gpu_task = NULL;
+        task = NULL;
+
+        gpu_task = (parsec_gpu_task_t*)parsec_fifo_pop( &(gpu_device->to_complete) );
+        if( gpu_task != NULL)
+        {
+            task = gpu_task->ec;
+            __parsec_complete_execution( es, task );
+            PARSEC_DEBUG_VERBOSE(4, parsec_gpu_output_stream,"GPU[%s]: gpu_task %p completed by co-manager %d at %s:%d", gpu_device->super.name,
+                                 gpu_task, es->th_id, __FILE__, __LINE__);
+            parsec_atomic_fetch_dec_int32( &(gpu_device->complete_mutex) );
+            parsec_list_push_back(gpu_tasks_to_free, (parsec_list_item_t*)gpu_task);
```

Comment:
Why do we have to accumulate task objects here? Can they not be free'd directly?

Reply:
Yes, thank you for pointing that out. i found a fix for this, it will be applied